### PR TITLE
feat: add ExecuteUntil to event dispatchers, add EventDispatcherThreadAware

### PR DESCRIPTION
**Pull request by richardapeters**

### CODE REVIEW

File: infra/event/test/TestEventDispatcherThreadAware.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                    done = true;
+                });
+        });
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
+        });
+
+    EXPECT_EQ(1, idleCount);
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcherThreadAware.cpp:44]:
```suggestion
    EXPECT_THAT(idleCount, testing::Eq(1));
```

File: infra/event/test/TestEventDispatcherWithWeakPtr.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    infra::EventDispatcherWithWeakPtr::Instance().Schedule([&]() {});
+
+    EXPECT_CALL(*this, Idle()).WillOnce(testing::Invoke([&]()
+        {
+            done = true;
+        }));
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcherWithWeakPtr.cpp:137]:
```suggestion
            return done;
```

File: infra/event/test/TestEventDispatcherWithWeakPtr.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    EXPECT_CALL(*this, RequestExecution());
+    infra::EventDispatcherWithWeakPtr::Instance().Schedule([&]()
+        {
+            done = true;
+        });
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcherWithWeakPtr.cpp:119]:
```suggestion
            return done;
```

File: infra/event/test/TestEventDispatcherWithWeakPtr.cpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            EXPECT_CALL(*this, RequestExecution());
+            infra::EventDispatcherWithWeakPtr::Instance().Schedule([&]()
+                {
+                    done = true;
+                });
+        }));
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcherWithWeakPtr.cpp:159]:
```suggestion
            return done;
```

File: infra/event/test/TestEventDispatcherThreadAware.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+            infra::EventDispatcher::Instance().Schedule([&]()
+                {
+                    done = true;
+                });
+        });
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcherThreadAware.cpp:41]:
```suggestion
            return done;
```

File: infra/event/test/TestEventDispatcher.cpp [hunk 6]
```diff
@@ -63,6 +60,75 @@ TEST_F(EventDispatcherTest, TestExecuteOneEvent)
...
+            EXPECT_CALL(*this, RequestExecution());
+            infra::EventDispatcher::Instance().Schedule([&]()
+                {
+                    done = true;
+                });
+        }));
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcher.cpp:128]:
```suggestion
            return done;
```

File: infra/event/test/TestEventDispatcher.cpp [hunk 7]
```diff
@@ -63,6 +60,75 @@ TEST_F(EventDispatcherTest, TestExecuteOneEvent)
...
+    infra::EventDispatcher::Instance().Schedule([&]() {});
+
+    EXPECT_CALL(*this, Idle()).WillOnce(testing::Invoke([&]()
+        {
+            done = true;
+        }));
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcher.cpp:106]:
```suggestion
            return done;
```

File: infra/event/test/TestEventDispatcher.cpp [hunk 8]
```diff
@@ -63,6 +60,75 @@ TEST_F(EventDispatcherTest, TestExecuteOneEvent)
...
+
+    EXPECT_CALL(*this, RequestExecution());
+    infra::EventDispatcher::Instance().Schedule([&]()
+        {
+            done = true;
+        });
+
+    ExecuteUntil([&]()
+        {
+            return done == true;
```
Comment (daantimmer) [inline infra/event/test/TestEventDispatcher.cpp:88]:
```suggestion
            return done;
```

File: infra/event/EventDispatcherWithWeakPtr.cpp [hunk 9]
```diff
@@ -46,6 +46,19 @@ namespace infra
...
     }
 
+    void EventDispatcherWithWeakPtrWorker::ExecuteUntil(const infra::Function<bool()>& predicate)
+    {
+        while (!predicate())
+        {
+            ExecuteAllActions();
+
+            if (predicate())
+                break;
```
Comment (daantimmer) [inline infra/event/EventDispatcherWithWeakPtr.cpp:56]:
```suggestion
                return;
```

File: infra/event/EventDispatcher.cpp [hunk 10]
```diff
@@ -46,6 +46,19 @@ namespace infra
...
     }
 
+    void EventDispatcherWorkerImpl::ExecuteUntil(const infra::Function<bool()>& predicate)
+    {
+        while (!predicate())
+        {
+            ExecuteAllActions();
+
+            if (predicate())
+                break;
```
Comment (daantimmer) [inline infra/event/EventDispatcher.cpp:56]:
```suggestion
                return;
```

**Review by daantimmer — COMMENTED** [review]
I think the current ExecuteUntil functions should take an templates Predicate. This allows the caller to provide a lambda with more captures than what infra:: Function allows.

Much like how the Subject/Observer Notify functions work.

File: infra/event/EventDispatcherWithWeakPtr.cpp (hunk 9 above)
Comment (richardapeters) [reply infra/event/EventDispatcherWithWeakPtr.cpp:56]: I actually prefer break over return. Post condition of the while-loop is that predicate() hold. Post condition has been found to hold, so we break from the loop.

**Comment by richardapeters** [comment]
> I think the current ExecuteUntil functions should take an templates Predicate. This allows the caller to provide a lambda with more captures than what infra:: Function allows.

Using infra::Function avoids having a method template, and is in line with how amp-embedded-infra-lib uses functions. I am still considering adding an interface on top of Run/ExecuteAllActions/ExecuteUntil somewhere in the future; that requires ExecuteUntil to not be a method template.

**Review by daantimmer — APPROVED** [review]

**Pull request merged**